Mathlib/LinearAlgebra/Reflection.lean

```diff
@@ -174,33 +174,47 @@ lemma Dual.eq_of_preReflection_mapsTo' [CharZero R] [NoZeroSMulDivisors R M]
 variable {y}
 variable {g : Dual R M}
 
+/-- Composite of reflections in "parallel" hyperplanes is a shear (special case). -/
+lemma reflection_reflection_iterate (hfx : f x = 2) (hgy : g y = 2) (hgxfy : f y * g x = 4)
+    (n : ℕ) : ((Module.reflection hgy).trans (Module.reflection hfx))^[n] y =
+      y + (n : ℤ) • (f y • x - (2 : ℤ) • y) := by
+  induction n with
+  | zero => simp
+  | succ n ih =>
+    have hz : ∀ z : M, f y • g x • z = 2 • 2 • z := by
+      intro z
+      rw [smul_smul, hgxfy, smul_smul, ← Nat.cast_smul_eq_nsmul R (2 * 2), Nat.cast_eq_ofNat]
+    simp only [iterate_succ', comp_apply, ih, map_add, LinearEquiv.trans_apply, sub_smul, map_smul,
+      reflection_apply_self, map_neg, reflection_apply, hz, two_smul, neg_sub, map_sub, smul_sub,
+      map_nsmul, map_zsmul, smul_neg, smul_add, Nat.cast_succ, mul_add, mul_one, add_smul, two_mul]
+    abel
+
+lemma infinite_range_of_reflection_reflection_iterate [NoZeroSMulDivisors ℤ M] (hfx : f x = 2)
+    (hgy : g y = 2) (hgxfy : f y * g x = 4) (hxy : ¬ f y • x = (2 : ℤ) • y) : Set.Infinite (range
+      (fun n => ((Module.reflection hgy).trans (Module.reflection hfx))^[n] y)) := by
+  refine infinite_range_of_injective fun n m hnm => ?_
+  simp only [reflection_reflection_iterate hfx hgy hgxfy] at hnm
+  rw [add_right_inj, ← sub_eq_zero, ← sub_smul, smul_eq_zero, sub_eq_zero, sub_eq_zero] at hnm
+  simpa using hnm.resolve_right hxy
```

Comment:
I think there is a tiny result worth factoring out here. To be precise I suggest putting these lemmas at the bottom of `Mathlib.Algebra.Module.Torsion`:
```lean
@[simp]
lemma infinite_range_add_smul_iff
    [AddCommGroup M] [Ring R] [Module R M] [Infinite R] [NoZeroSMulDivisors R M] (x y : M) :
    (Set.range <| fun r : R ↦ x + r • y).Infinite ↔ y ≠ 0 := by
  refine ⟨fun h hy ↦ by simp [hy] at h, fun h ↦ Set.infinite_range_of_injective fun r s hrs ↦ ?_⟩
  rw [add_right_inj] at hrs
  exact smul_left_injective _ h hrs

@[simp]
lemma infinite_range_add_nsmul_iff
    [AddCommGroup M] [NoZeroSMulDivisors ℤ M] (x y : M) :
    (Set.range <| fun n : ℕ ↦ x + n • y).Infinite ↔ y ≠ 0 := by
  refine ⟨fun h hy ↦ by simp [hy] at h, fun h ↦ Set.infinite_range_of_injective fun r s hrs ↦ ?_⟩
  rw [add_right_inj, ← natCast_zsmul, ← natCast_zsmul] at hrs
  simpa using smul_left_injective _ h hrs
```
so that we can shorten the proof here (and also upgrade it to an iff) as follows:
```suggestion
lemma infinite_range_reflection_reflection_iterate_iff [NoZeroSMulDivisors ℤ M]
    (hfx : f x = 2) (hgy : g y = 2) (hgxfy : f y * g x = 4) :
    (range <| fun n ↦ ((reflection hgy).trans (reflection hfx))^[n] y).Infinite ↔
    f y • x ≠ (2 : R) • y := by
  simp only [reflection_reflection_iterate hfx hgy hgxfy, infinite_range_add_nsmul_iff, sub_ne_zero]
```
I was 50-50 on whether we should state these results or their contrapositive but given your ultimate application I was persuaded this was the most convenient.